cmd/ecc/join.go

```diff
@@ -0,0 +1,330 @@
+// erasure coding reconstruction.
+func (r *JoinedReaders) Read(b *Buffer) error {
+	if len(r.src) != len(b.shards) {
+		panic("ecc: number of data sources does not match the number of shards")
+	}
+	if r.rErr != nil {
+		return r.rErr
+	}
+
+	b.Reset()
+	results, next, err := r.spawn(b)
+	if err != nil {
+		r.rErr = err
+		return r.rErr
+	}
+
+	// Mark all shards as missing.
+	// When a read from a data source succeeds
+	// the particular shard will get marked
+	// as available again.
+	for i := range b.shards {
+		b.shards[i] = b.shards[i][:0]
+	}
+
+	return r.join(b, next, results)
+}
+
+type readResult struct {
+	ID  int   // The ID / index of the shard / data source
+	N   int   // The number of bytes read from the data source
+	Err error // Any error that happened during reading
+}
+
+// spawn creates just as many concurrent go routines as necessary to read
+// enough shards to reconstruct the content data. If all data shards are
+// available each go routine tries to read a data shard.
+// It returns a channel which hold the read results once they are available.
+// The returned next index points to the next data source from which a go
+// routine should read if any of the spawned reads fails.
+func (r *JoinedReaders) spawn(b *Buffer) (results chan readResult, next int, err error) {
+	results = make(chan readResult, len(r.src))
+	var spawned int
+	for spawned < len(b.data) && next < len(b.shards) {
+		if r.err[next] == nil {
+			go readFrom(r.src[next], next, r.offset, b.shards[next], results)
+			spawned++
+		}
+		next++
+	}
+	// If we could not spwan a single go routine we should return an error
+	// (or panic) since a join would block forever since the results channel
+	// would never receive a result. However, that error should never happen
+	// since there should be at least 1 (actually `len(b.data)`) non-nil data
+	// sources. Otherwise, we never had read quorum in the first place and
+	// should have failed somewhere higher up the stack.
+	// TODO(aead): Consider a panic.
+	if spawned == 0 {
+		err = ErrReadQuorum
+	}
+	return
+}
+
+// join joins all go routines created by spawn(). Therefore, it reads
+// from the results channel until it has received enough shards. If
+// one read fails - i.e. returns an non-EOF error - it re-spawns another
+// go routine reading from the next data source.
+func (r *JoinedReaders) join(b *Buffer, next int, results chan readResult) error {
+	var (
+		n      int // Number of bytes read in total
+		nReads int // Number of reads that succeeded
+		nEOFs  int // Number of cases read returned an EOF error
+	)
+	for result := range results {
+		i, nn, err := result.ID, result.N, result.Err
+
+		// In case the i-th read fails we check whether
+		// performing another read "makes sense". (If we
+		// have encountered more the `N = parity` errors
+		// we won't be able to reconstruct the data - so
+		// we return ErrReadQuorum).
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
```

Comment:
Ignoring `io.ErrUnexpectedEOF` (or equalling it to `io.EOF`) seems quite dangerous. What is the reason for ignoring it?

You seems to be making assumptions about what it?

```suggestion
		if err != nil && !errors.Is(err, io.EOF) && !errors.Is(err, io.ErrUnexpectedEOF) {
```

Reply:
This is the error returned by the `io.ReadAt` call of the spawned go routine. So, the assumption is that the data source implementation tries to read as many bytes as possible into the buffer (See: `io.ReadFull`). So `io.ErrUnexpectedEOF` indicates that the caller read `n > 0` bytes but was not able to fill the buffer due to an `io.EOF`.

We can only remove this if we require that the `ReaderAt` implementation never returns `io.ErrUnexpectedEOF`
in case of `n > 0` and an encountered `io.EOF`.

Reply:
The problem is that this cannot be distinguished from an `io.ErrUnexpectedEOF` which could be returned for other reasons. I assume we don't have any way of knowing which sizes to expect?

Maybe switch to `io.ReadAtLeast` with n=0? That way upstream errors will be preserved.

Reply:
That cannot happen due to the fact that the underlying `io.ReaderAt` is a bitrot verification reader which won't accept an expected EOF since the checksums won't match. However, by requiring that the bitrot verification reader never returns `io.ErrUnexpectedEOF` but `io.EOF` we can get rid of the `io.ErrUnexpectedEOF` check here.